packages/docid/src/index.ts

```diff
@@ -0,0 +1,210 @@
+   * @param {string}             [multibaseName = 'base36']
+   *
+   * @example
+   * new DocId(<docType>, <CID>|<cidStr>)
+   * new DocId(<docType>, <CID>|<cidStr>, <subnet>, <multibaseName>)
+   */
+
+  private _docType: number 
+  private _subnet: number
+  private _cid: CID
+  private _multibaseName: string
+
+  constructor (doctype: string | number = 'tile', cid: CID | string, subnet: string | number = 'devnet', multibaseName: string = 'base36') {
+    this._docType = (typeof cid === 'string') ? doctypes[doctype] : doctype
+     // if (!doctype) throw new Error('constructor: doctype required')
+    this._subnet = (typeof cid === 'string')? subnets[subnet] : subnet
+    this._multibaseName = multibaseName 
+    this._cid = (typeof cid === 'string') ? new CID(cid) : cid
+      // if (!cid) throw new Error('constructor: cid required')
+  }
+
+  static fromBytes(bytes: Uint8Array, multibaseName: string): DocId {
+    const subnet = varint.decode(bytes)
+    bytes = bytes.slice(varint.decode.bytes) 
+    const docType = varint.decode(bytes)
+    bytes = bytes.slice(varint.decode.bytes) 
+    const cid = new CID(bytes)
+    return new DocId(docType, cid, subnet, multibaseName)
+  }
+
+  static fromString(docId: string): DocId {
+    const multibaseName = multibase.isEncoded(docId)
+    const bytes = multibase.decode(docId).slice(2)
+    return DocId.fromBytes(bytes, multibaseName)
+  }
+
+  /**
+   * Get doc type code
+   *
+   * @returns {number}
+   * @readonly
+   */
+  get docType (): number {
+    return this._docType
+  }
+
+  /**
+   * Get doc type name
+   *
+   * @returns {string}
+   * @readonly
+   */
+  get docTypeName (): string {
+    const name = getKey(doctypes, this._docType)
+    if (!name) throw new Error('docTypeName: no registered name available')
+    return name 
+  }
+
+  /**
+   * Get CID
+   *
+   * @returns {CID}
+   * @readonly
+   */
+  get cid (): CID {
+    return this._cid
+  }
+
+  /**
+   * Get subnet code
+   *
+   * @returns {number}
+   * @readonly
+   */
+  get subnet (): number {
+    return this._subnet
+  }
+
+  /**
+   * Get subnet name 
+   *
+   * @returns {string}
+   * @readonly
+   */
+  get subnetName (): string {
+    const name = getKey(doctypes, this._subnet)
+    if (!name) throw new Error('subnetName: no registered name available')
+    return name 
+  }
+
+  /**
+   * Get multibase name string
+   *
+   * @returns {string}
+   * @readonly
+   */
+  get multibaseName (): string {
+    return this._multibaseName
+  }
+
+   /**
+   * Get multihash bytes of DocId
+   *
+   * @returns {Uint8Array}
+   * @readonly
+   */
+  get multihash (): Uint8Array {
+    return this._cid.multihash
+  }
+
+   /**
+   * Get CID codec name string
+   *
+   * @returns {string}
+   * @readonly
+   */
+  get codec (): string {
+    return this._cid.codec
+  }
+
+  /**
+   * Get bytes of DocId
+   *
+   * @returns {Uint8Array}
+   * @readonly
+   */
+  get bytes (): Uint8Array {
```

Comment:
Could we cache this value please? Ex:
```js
get bytes (): Uint8Array {
  if (this._bytes == null) {
    ...
    this._bytes = uint8ArrayConcat(...)
  }
  return this._bytes
}
```

This way if the consumer uses `docid.bytes` multiple times, the value is cached rather than being computed every time.